homeassistant/components/recorder/statistics.py

```diff
@@ -1597,17 +1639,59 @@ def _statistics_during_period_with_session(
         types,
     )
 
-    # Return statistics combined with metadata
-    if period not in ("day", "week", "month"):
-        return result
-
     if period == "day":
-        return _reduce_statistics_per_day(result, types)
+        result = _reduce_statistics_per_day(result, types)
 
     if period == "week":
-        return _reduce_statistics_per_week(result, types)
+        result = _reduce_statistics_per_week(result, types)
 
-    return _reduce_statistics_per_month(result, types)
+    if period == "month":
+        result = _reduce_statistics_per_month(result, types)
+
+    if "change" in _types:
+        drop_sum = "sum" not in _types
+        prev_sums = {}
+        if tmp := _statistics_at_time(
+            session,
+            {metadata[statistic_id][0] for statistic_id in result},
+            table,
+            start_time,
+            {"sum"},
+        ):
+            _metadata = dict(metadata.values())
+            for row in tmp:
+                metadata_by_id = _metadata[row.metadata_id]
+                statistic_id = metadata_by_id["statistic_id"]
+
+                state_unit = unit = metadata_by_id["unit_of_measurement"]
+                if state := hass.states.get(statistic_id):
+                    state_unit = state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
+                convert = _get_statistic_to_display_unit_converter(
+                    unit, state_unit, units
+                )
+
+                if convert is not None:
+                    prev_sums[statistic_id] = convert(row.sum)
+                else:
+                    prev_sums[statistic_id] = row.sum
+
+        for statistic_id, rows in result.items():
+            prev_sum = prev_sums.get(statistic_id) or 0
+            for statistics_row in rows:
+                if "sum" not in statistics_row:
+                    continue
+                if drop_sum:
+                    _sum = statistics_row.pop("sum")
+                else:
+                    _sum = statistics_row["sum"]
+                if _sum is None:
+                    statistics_row["change"] = None
+                    continue
+                statistics_row["change"] = _sum - prev_sum
+                prev_sum = _sum
```

Comment:
Maybe avoid one `branch` each loop?

```suggestion
            if drop_sum:
                _getter = statistics_row.pop
            else:
                _getter = statistics_row.get
            for statistics_row in rows:
                if "sum" not in statistics_row:
                    continue
                if (_sum := _getter("sum", None)) is None:
                    statistics_row["change"] = None
                    continue
                statistics_row["change"] = _sum - prev_sum
                prev_sum = _sum
```

Reply:
`_getter` won't work because `statistics_row` changes every loop.

Reply:
You could always split it into two loops but that makes it a bit WET

Reply:
I think we should profile first since this happens after reducing hourly data points to daily / weekly / monthly.

Reply:
No need to spend time optimizing this. All the time is spent doing the db query